src/detect-engine-threshold.c

```diff
@@ -39,65 +39,414 @@
 #include "detect.h"
 #include "flow.h"
 
-#include "host.h"
-#include "host-storage.h"
-
-#include "ippair.h"
-#include "ippair-storage.h"
-
 #include "detect-parse.h"
-#include "detect-engine-sigorder.h"
-
-#include "detect-engine-siggroup.h"
-#include "detect-engine-address.h"
-#include "detect-engine-port.h"
-#include "detect-engine-mpm.h"
-#include "detect-engine-iponly.h"
-
 #include "detect-engine.h"
 #include "detect-engine-threshold.h"
+#include "detect-engine-address.h"
+#include "detect-engine-address-ipv6.h"
 
-#include "detect-content.h"
-#include "detect-uricontent.h"
-
-#include "util-hash.h"
+#include "util-misc.h"
 #include "util-time.h"
 #include "util-error.h"
 #include "util-debug.h"
+#include "action-globals.h"
+#include "util-validate.h"
 
-#include "util-var-name.h"
-#include "tm-threads.h"
+#include "util-hash.h"
+#include "util-thash.h"
+#include "util-hash-lookup3.h"
 
-#include "action-globals.h"
+struct Thresholds {
+    THashTableContext *thash;
+} ctx;
 
-static HostStorageId host_threshold_id = { .id = -1 };     /**< host storage id for thresholds */
-static IPPairStorageId ippair_threshold_id = { .id = -1 }; /**< ip pair storage id for thresholds */
+static int ThresholdsInit(struct Thresholds *t);
+static void ThresholdsDestroy(struct Thresholds *t);
 
-HostStorageId ThresholdHostStorageId(void)
+void ThresholdInit(void)
 {
-    return host_threshold_id;
+    ThresholdsInit(&ctx);
 }
 
-void ThresholdInit(void)
+void ThresholdDestroy(void)
+{
+    ThresholdsDestroy(&ctx);
+}
+
+#define SID    0
+#define GID    1
+#define REV    2
```

Comment:
Why did we add rev since at any given time, except for rule reload we shouldn't have duplicate SIDs? Just another source of data for the hash?

Reply:
rev can change with rule reloads, and this data structure is global. In the existing implementation this is messy. Per rule thresholds are in the de_ctx, but the per host/ippair are also global (host/ippair table). There we don't account for rev, which means a changed rule might not be handled correctly.